tests/ui/associated-types/substs-ppaux.normal.stderr

```diff
@@ -70,13 +70,12 @@ help: use parentheses to call this function
 LL |     let x: () = foo::<'static>();
    |                               ++
 
-error[E0277]: the size for values of type `str` cannot be known at compilation time
+error[E0277]: the trait bound `str: Foo<'_, '_, u8>` is not satisfied
   --> $DIR/substs-ppaux.rs:49:6
    |
 LL |     <str as Foo<u8>>::bar;
-   |      ^^^ doesn't have a size known at compile-time
+   |      ^^^ the trait `Sized` is not implemented for `str`, which is required by `str: Foo<'_, '_, u8>`
```

Comment:
I haven't read the implementation or the most recent Zulip comment and just saw this stderr diff, so sorry if this has been discussed already.

Is this intentional? Shouldn't we still use the `[rustc_on_unimplemented]`/`#[diagnostic::on_unimplemented]` message here despite the derived/root obligation changes?

Reply:
It is somewhat intentional. When you have `T: Sized` then you want `T` to have a compile time known size. When you have `T: Foo, Foo: Sized`, I think it makes sense to go back to "`T` does not implement `Sized`", it's a subtle difference of "what does the user care about".

Reply:
Also, I do want to do a full refactor of this logic, too many ad-hoc hacks have piled beyond maintainability (in no small part due to me), but now that we know what the *output* we want to keep and change is, it will be easier to clean this up in a way that makes sense. This PR is about putting the output piece closer to that "ideal end state" in anticipation of that work.